### PR TITLE
feat: project group access endpoints

### PR DESCRIPTION
**Pull request by IrakliJani**
related to: https://github.com/lightdash/lightdash/issues/8499

adds:
- `project_group_access` table creation migration
- entity for `projectGroupAccess`
- `addProjectAccess` and `removeProjectAccess` to group model
- `addProjectAccess` and `removeProjectAccess` to group service with permission checks 
- `POST`(add) and `DELETE`(remove) routes: `/v1/groups/{groupUuid}/projects/${{projectUuid}`
- e2e tests for newly added api. routes

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/backend/src/database/migrations/20240109124309_project_group_access.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                    .inTable(ORGANIZATIONS_TABLE_NAME)
+                    .onDelete('CASCADE');
+                table
+                    .text('role')
+                    .references('role')
+                    .inTable(PROJECT_MEMBERSHIP_ROLES_TABLE_NAME)
+                    .notNullable()
+                    .onDelete('RESTRICT')
+                    .defaultTo('viewer');
+                table.unique(['project_id', 'group_uuid', 'organization_id']);
```
Comment (owlas) [inline packages/backend/src/database/migrations/20240109124309_project_group_access.ts:39]: I don't think we need organization_id here, project_id is unique on it's own

File: packages/backend/src/database/migrations/20240109124309_project_group_access.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                    .references('project_id')
+                    .inTable(PROJECTS_TABLE_NAME)
+                    .onDelete('CASCADE');
+                table
+                    .uuid('group_uuid')
+                    .notNullable()
+                    .references('group_uuid')
+                    .inTable(GROUPS_TABLE_NAME)
+                    .onDelete('CASCADE');
+                table
```
Comment (owlas) [inline packages/backend/src/database/migrations/20240109124309_project_group_access.ts:26]: remove

File: packages/backend/src/database/migrations/20240109124309_project_group_access.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                    .integer('organization_id')
+                    .notNullable()
+                    .references('organization_id')
+                    .inTable(ORGANIZATIONS_TABLE_NAME)
+                    .onDelete('CASCADE');
+                table
+                    .text('role')
+                    .references('role')
+                    .inTable(PROJECT_MEMBERSHIP_ROLES_TABLE_NAME)
```
Comment (owlas) [inline packages/backend/src/database/migrations/20240109124309_project_group_access.ts:35]: great work!

File: packages/backend/src/database/migrations/20240109124309_project_group_access.ts (hunk 1 above)
Comment (IrakliJani) [reply packages/backend/src/database/migrations/20240109124309_project_group_access.ts]: done

File: packages/e2e/cypress/e2e/api/groups.cy.ts [hunk 4]
```diff
@@ -192,4 +193,52 @@ describe('Groups API', () => {
             });
         });
     });
+
+    describe('Group Project access API', () => {
```
Comment (ZeRego) [inline packages/e2e/cypress/e2e/api/groups.cy.ts:197]: nice 🙌

File: packages/backend/src/database/migrations/20240109124309_project_group_access.ts (hunk 2 above)
Comment (IrakliJani) [reply packages/backend/src/database/migrations/20240109124309_project_group_access.ts:25]: ✔️

File: packages/backend/src/database/migrations/20240109124309_project_group_access.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+const PROJECT_MEMBERSHIP_ROLES_TABLE_NAME = 'project_membership_roles';
+const PROJECT_GROUP_ACCESS_TABLE_NAME = 'project_group_access';
+
+export async function up(knex: Knex): Promise<void> {
+    if (!(await knex.schema.hasTable(PROJECT_GROUP_ACCESS_TABLE_NAME))) {
+        await knex.schema.createTable(
+            PROJECT_GROUP_ACCESS_TABLE_NAME,
+            (table) => {
+                table
+                    .integer('project_id')
```
Comment (ZeRego) [inline packages/backend/src/database/migrations/20240109124309_project_group_access.ts:14]: We used ids in the past for foreign keys but new tables use uuids to simplify the queries in models.

Comment (ZeRego) [reply packages/backend/src/database/migrations/20240109124309_project_group_access.ts:14]: With this change we can stop using knex.raw in favour or other functions.

```
const query = this.database.raw(
            `
            DELETE FROM ${ProjectGroupAccessTableName}
            WHERE project_id = (
                SELECT project_id
                FROM ${ProjectTableName}
                WHERE project_uuid = :projectUuid
            )
            AND group_uuid = :groupUuid
            RETURNING *
            `,
            { projectUuid, groupUuid },
        );
```
With 
```
const query = this.database.delete().where('project_uuid', projectUuid);
```

And 

```
     const query = this.database.raw<{
            rows: DBProjectGroupAccess[];
        }>(
            `
            INSERT INTO ${ProjectGroupAccessTableName} (project_id, group_uuid, role)
            SELECT ${ProjectTableName}.project_id, ${GroupTableName}.group_uuid, :role
            FROM ${ProjectTableName}
            INNER JOIN ${GroupTableName} ON ${ProjectTableName}.organization_id = ${GroupTableName}.organization_id
            WHERE ${ProjectTableName}.project_uuid = :projectUuid
            AND ${GroupTableName}.group_uuid = :groupUuid
            AND ${ProjectTableName}.organization_id = ${GroupTableName}.organization_id
            ON CONFLICT DO NOTHING
            RETURNING *
            `,
            { projectUuid, groupUuid, role },
        )
```

With 

```
   const query = this.database..insert( { projectUuid, groupUuid, role })
```

Comment (IrakliJani) [reply packages/backend/src/database/migrations/20240109124309_project_group_access.ts:14]: @ZeRego yeah, that makes sense! I'll make the changes 🙌 

```
const query = this.database.insert( { projectUuid, groupUuid, role })
```

this is simple but lacks db integrity check where group and project should belong to the same org. I'll try the KNEX way of that query

File: packages/frontend/src/features/projectGroupAccess/components/ProjectGroupAccessModal.tsx [hunk 6]
```diff
@@ -71,6 +76,45 @@ const ProjectGroupAccessModal: FC<ProjectGroupAccessModalProps> = ({
...
+                groupUuid: groups[0].uuid,
+            });
+
+            const { data: data3 } = await refetchProjectAccesses();
+            console.log('after removing project accesses', data3);
+        }
+
+        demo();
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [groups]);
```
Comment (ZeRego) [inline packages/frontend/src/features/projectGroupAccess/components/ProjectGroupAccessModal.tsx:116]: Do we want to remove this test code before merge?

Comment (IrakliJani) [reply packages/frontend/src/features/projectGroupAccess/components/ProjectGroupAccessModal.tsx]: removed

**Review by ZeRego — APPROVED** [review]

**Pull request merged**